--- a/onnxruntime/core/providers/cuda/cuda_common.h
+++ b/onnxruntime/core/providers/cuda/cuda_common.h
@@ -5,12 +5,6 @@
 
 #ifndef BUILD_CUDA_EP_AS_PLUGIN
 
-// The following three lines were copied from ABSL
-// cutlass needs them, because cutlass uses "and"/"or" keywords
-#ifdef __cplusplus
-#include <ciso646>
-#endif
-
 #if defined(ENABLE_FP4) && !defined(DISABLE_FLOAT4_TYPES)
 
 #if defined(_MSC_VER)
--- a/onnxruntime/core/optimizer/matmul_nbits_fusion.h
+++ b/onnxruntime/core/optimizer/matmul_nbits_fusion.h
@@ -15,7 +15,7 @@
   MatMulNBitsFusion(const InlinedHashSet<std::string_view>& compatible_eps = {},
                     const SatApplyContextVariant& apply_context = {});
 
-  SelectorActionRegistry CreateSelectorActionRegistry() const;
+  static SelectorActionRegistry CreateSelectorActionRegistry();
 };
 
 }  // namespace onnxruntime
--- a/onnxruntime/core/optimizer/matmul_nbits_fusion.cc
+++ b/onnxruntime/core/optimizer/matmul_nbits_fusion.cc
@@ -123,7 +123,7 @@
 
 }  // namespace
 
-SelectorActionRegistry MatMulNBitsFusion::CreateSelectorActionRegistry() const {
+SelectorActionRegistry MatMulNBitsFusion::CreateSelectorActionRegistry() {
   SelectorActionRegistry registry{};
 
   BiasFusionRule(registry);
